### PR TITLE
filters: wire Android invocation path

### PR DESCRIPTION
**Pull request by goaway**
Description: Adds invocation path for Android platform-bridged filters, covering request/response headers, data, and trailers. This change does not provide for state modification via the return path.
Risk Level: Moderate
Testing: Local and CI

Signed-off-by: Mike Schore <mike.schore@gmail.com>

### CODE REVIEW

File: examples/kotlin/hello_world/MainActivity.kt [hunk 1]
```diff
@@ -36,7 +37,9 @@ class MainActivity : Activity() {
     super.onCreate(savedInstanceState)
     setContentView(R.layout.activity_main)
 
-    streamClient = AndroidStreamClientBuilder(application).build()
+    streamClient = AndroidStreamClientBuilder(application)
+      .addFilter({ DemoFilter() })
```
Comment (buildbreaker) [inline examples/kotlin/hello_world/MainActivity.kt:41]:
```suggestion
      .addFilter { DemoFilter() }
```

You don't have to add the parens for this case

File: library/java/src/io/envoyproxy/envoymobile/engine/JvmFilterContext.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+import java.nio.ByteBuffer;
+import java.util.Map;
+
+import io.envoyproxy.envoymobile.engine.types.EnvoyHTTPFilter;
+
+/**
+ * Wrapper class for EnvoyHTTPFilter for receiving JNI calls.
+ */
+class JvmFilterContext {
```
Comment (buildbreaker) [inline library/java/src/io/envoyproxy/envoymobile/engine/JvmFilterContext.java:11]: Is it possible for us to add unit tests to just ensure the the filters are called appropriately?

Comment (goaway) [reply library/java/src/io/envoyproxy/envoymobile/engine/JvmFilterContext.java:11]: https://github.com/lyft/envoy-mobile/pull/1024

Comment (buildbreaker) [reply library/java/src/io/envoyproxy/envoymobile/engine/JvmFilterContext.java:11]: Could we also have a `JvmFilterContext` unit test which helps us with exercising the interaction between the `JvmBridgeUtility` and `EnvoyHTTPFilter`?

Comment (goaway) [reply library/java/src/io/envoyproxy/envoymobile/engine/JvmFilterContext.java:11]: Stepping back for a second, I'm honestly skeptical of the value of **unit** testing functions that purely perform delegation. There's no real black box logic to test; the function makes another call, and you implement that call on one hand and then assert you've implemented it on the other. In my opinion, integration testing is a much better mechanism for testing for correct wiring/coupling of layers.

Translation on the other hand, I think merits unit testing - which is partly why I've extracted logic into the `JvmBridgeUtility` (and also the `JniBridgeUtility` in the upcoming PR).

Comment (buildbreaker) [reply library/java/src/io/envoyproxy/envoymobile/engine/JvmFilterContext.java:11]: That's fair. I would like to capture and exercise the delegation path somewhere.

I assume you have in mind an integration test which would exercise this?

**Comment by buildbreaker** [comment]
Looks good. Let me know once the formatter is green and I can approve

File: library/java/src/io/envoyproxy/envoymobile/engine/JvmFilterContext.java (hunk 2 above)
Comment (goaway) [reply library/java/src/io/envoyproxy/envoymobile/engine/JvmFilterContext.java:11]: Yes, I'd like to cover the paths through the layer wiring using the AssertionFilter that @junr03 has implemented.

**Comment by goaway** [comment]
Tested example Swift app (which is failing in CI, but should be entirely unrelated anyways) locally.

**Pull request merged**